oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/DataModelService.kt

```diff
@@ -1,92 +1,176 @@
+    fun closeSim(name: String) {
+        require(name in simByName)
+        try {
+            simByName.getValue(name).input.onCompleted()
+        } finally {
+            removeSim(name)
+        }
+    }
+
+    fun removeSim(name: String) {
+        require(name in simByName)
+        simByName -= name
+        eventBus.post(PluginUnRegisteredEvent(name))
+        updateIssueList()
     }
 
-    override fun updateSim(newNode: Simulation) {
-        require(simByName.containsKey(newNode.name))
+    fun updateSim(newNode: Simulation) {
+        require(newNode.name in simByName)
         simByName += newNode.name to newNode
-        rebindView()
+        eventBus.post(PluginUpdatedEvent())
+        updateIssueList()
     }
 
-    private fun addNodeToView(name: String) {
-        GlobalScope.launch(Dispatchers.JavaFx) {
-            viewData.nodes.add(name)
+    fun addConfiguration(name: String) {
+        if (name in proxyByName) {
+            return
         }
+        proxyByName += name to Proxy(name)
+        eventBus.post(ProxyAddedEvent(name))
+        updateIssueList()
     }
 
-    private fun removeNodeFromView(name: String) {
-        GlobalScope.launch(Dispatchers.JavaFx) {
-            viewData.nodes.remove(name)
+    fun updateConfiguration(name: String) {
+        if (name !in proxyByName) {
+            return
+        }
+        val proxy = proxyByName[name]
+        val sim = simByName[name]
+        if (proxy == null || sim == null) {
+            TODO()
+        } else {
+            proxyByName += name to Proxy(name, sim.inputs, sim.outputs, proxy.timeOut)
         }
+
+        eventBus.post(ProxyUpdatedEvent())
+        updateIssueList()
     }
 
-    private fun removeAllNodesFromView() {
-        GlobalScope.launch(Dispatchers.JavaFx) {
-            viewData.nodes.clear()
+    fun updateConfiguration(newProxy: Proxy) {
+        if (newProxy.name !in proxyByName) {
+            return
         }
+
+        proxyByName += newProxy.name to newProxy
+        updateIssueList()
     }
 
-    private fun rebindView() {
-        GlobalScope.launch(Dispatchers.JavaFx) {
-            updateCallback.invoke()
+    fun removeConfiguration(name: String) {
+        if (name !in proxyByName) {
+            return
         }
+        proxyByName -= name
+        eventBus.post(ProxyRemovedEvent(name))
+        updateIssueList()
     }
 
-}
\ No newline at end of file
+
+    fun findIssue(): List<String> {
+        var issues = emptyList<String>()
+        val noMatchingSim = proxyByName.keys.filter { it !in simByName.keys }
+        if (noMatchingSim.isNotEmpty()) {
+            noMatchingSim.forEach {
+                issues += "Missing Simulation: Proxy setup \"$it\" reference to a missing simulation \"${proxyByName.getValue(it).name}\""
+            }
+        }
+
+        val noMatchingProxy = simByName.keys.filter { it !in proxyByName.keys }
+        if (noMatchingProxy.isNotEmpty()) {
+            noMatchingProxy.forEach {
```

Comment:
why do you need an `isNotEmpty` check?

Reply:
think you can maybe express this with a `map` query?

Reply:
It is kind of defined but not used problem. In may well be a warning but I don't have the concept introduced yet.